src/components/v5/common/Fields/Datepicker/partials/DatepickerCustomHeader/DatepickerCustomHeader.tsx

```diff
@@ -0,0 +1,141 @@
+import { CaretLeft, CaretRight, X } from '@phosphor-icons/react';
+import clsx from 'clsx';
+import format from 'date-fns/format';
+import getYear from 'date-fns/getYear';
+import isMatch from 'date-fns/isMatch';
+import parse from 'date-fns/parse';
+import React, {
+  type ChangeEventHandler,
+  type FC,
+  useEffect,
+  useState,
+} from 'react';
+
+import { formatText } from '~utils/intl.ts';
+import { range } from '~utils/lodash.ts';
+import InputBase from '~v5/common/Fields/InputBase/index.ts';
+
+import DatepickerYearDropdown from '../DatepickerYearDropdown/index.ts';
+
+import { type DatepickerCustomHeaderProps } from './types.ts';
+
+const years = range(1990, getYear(new Date()) + 1, 1);
+
+const DatepickerCustomHeader: FC<DatepickerCustomHeaderProps> = ({
+  date,
+  changeYear,
+  monthDate,
+  decreaseMonth,
+  increaseMonth,
+  prevMonthButtonDisabled,
+  nextMonthButtonDisabled,
+  startDate,
+  endDate,
+  dateFormat,
+  setStartDate,
+  setEndDate,
+  onClose,
+}) => {
+  const [startDateText, setStartDateText] = useState<string>('');
+  const [endDateText, setEndDateText] = useState<string>('');
+  const navigationButtonClassName =
+    'flex justify-center items-center p-2 text-gray-500 transition sm:hover:text-blue-400';
+
+  const makeChangeHandler =
+    (
+      setDateValue: (dateParam: Date) => void,
+      setTextValue: (dateText: string) => void,
+    ): ChangeEventHandler<HTMLInputElement> =>
+    (event) => {
+      const { value } = event.target;
+
+      setTextValue(value);
+
+      if (!isMatch(value, dateFormat)) {
+        return;
+      }
+
+      const parsedDate = parse(value, dateFormat, new Date());
+
+      setDateValue(parsedDate);
+    };
+
+  useEffect(() => {
+    if (startDate) {
+      setStartDateText(format(startDate, dateFormat));
+    }
+
+    if (endDate) {
+      setEndDateText(format(endDate, dateFormat));
+    }
+  }, [startDate, setStartDate, endDate, dateFormat]);
+
+  return (
+    <div className="w-full pt-5 pb-3 px-4 flex flex-col gap-3">
+      {onClose && (
+        <div className="w-full flex justify-end">
+          <button
+            type="button"
+            onClick={onClose}
+            className={navigationButtonClassName}
+          >
+            <X size={16} className="text-inherit" />
+          </button>
+        </div>
+      )}
+      <div className="w-full flex justify-between items-center gap-2">
+        <button
+          type="button"
+          onClick={decreaseMonth}
+          disabled={prevMonthButtonDisabled}
+          className={navigationButtonClassName}
+        >
+          <CaretLeft size={16} className="text-inherit" />
+        </button>
+        <DatepickerYearDropdown
+          years={years}
+          selectedYear={getYear(date)}
+          selectedMonth={monthDate.toLocaleString('en-US', {
```

Comment:
To some degree this should be tied into `i18n` as internationalization and languages should also change the numbers, time and date formats to they're local regional equivalent.

But I guess that's a larger discussion, so might want to hold off for now.